grails-app/domain/org/pih/warehouse/shipping/Shipment.groovy

```diff
@@ -196,12 +196,12 @@ class Shipment implements Comparable, Serializable, Historizable {
             })?.unique({ a, b -> a <=> b })?.size() == referenceNumbers?.size()
         })
 
-        // a shipment can't have two events with the same event code (this should be the case for the current event codes: CREATED, SHIPPED, RECEIVED)
-        // we may want to change this in the future?
+        // a shipment can't have two system events with the same event code
         events(validator: { events ->
-            events?.collect({
-                it.eventType?.eventCode
-            })?.unique({ a, b -> a <=> b })?.size() == events?.size()
+            List<Event> systemEvents = events.toList().findAll {
```

Comment:
a titbit, that you probably didn't have to parse it to list, since you have the `.findAll` method available also for the Set.

Reply:
When I am removing the `.toList()` I am getting:
```
Cannot cast object '[Shipped|fr:Exp Rwinkwavu Pharmacy Warehouse on 2024-08-23 16:02:22.0]' with class 'java.util.TreeSet' to class 'java.util.List' due to: groovy.lang.GroovyRuntimeException: Could not find matching constructor for: java.util.List(org.pih.warehouse.core.Event)
```

Reply:
Isn’t it due to the fact that you declared the returned type as List<Event>?